CodePrinter func rewrite precedence bug fix

mgreminger opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
I searched but didn't find any related issues.

#### Brief description of what is fixed or changed
The `NumPyPrinter` relies on the `_print_Function` method of the  `CodePrinter` class to rewrite functions that don't exist in NumPy (such as the `cot` and `sec` functions). For some functions, the rewriting process results in expressions that are more complex than a single function call in the target code (for example, `sec` gets rewritten as `cos**-1`). This can cause issues when the function occurs in a larger expression since it may change the precedence of the final expression.   

Example where the current behavior generates code that is incorrect because of precedence rules :
```python
In [1]: import sympy as sp

In [2]: from sympy.printing.numpy import NumPyPrinter

In [3]: x,y,z = sp.symbols('x,y,z')

In [4]: NumPyPrinter().doprint(sp.sec(x)**2)
Out[4]: 'numpy.cos(x)**(-1.0)**2'
```

Note that the generated code is equivalent to `cos(x)**1.0` instead of the intended `cos(x)**-2` because of the right associativity of the `**` operator.

The [precedence rules of Python](https://docs.python.org/3/reference/expressions.html#operator-precedence) place function calls and parenthesized expressions at about the same precedence level with parenthesized expressions being slightly higher than function calls. Because of this,  the safest approach to fix this problem would be to place all rewritten functions in parenthesis. However, in many cases, where the result of the rewrite is itself a function, the parenthesis are unnecessary.  To avoid unnecessary parenthesis in the generated code, I modified the `_print_Function` method to only add the parenthesis if the call to `rewrite` results in a top level function that is `Add`, `Mul`, or `Pow`. As far as I can tell, these are the only cases that would result in the generated code using Python operators directly, which leads to the possibility of the unintended change in precedence.  However, there could be corner cases I'm missing. If it's preferable to take the safer option of always wrapping the expression resulting from the rewrite in parenthesis, let me know and I'll make the change.

New Behavior:
```python
In [1]: import sympy as sp

In [2]: from sympy.printing.numpy import NumPyPrinter

In [3]: x,y,z = sp.symbols('x,y,z')

In [4]: NumPyPrinter().doprint(sp.sec(x)**2)
Out[4]: '(numpy.cos(x)**(-1.0))**2'
```

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* printing
  * Fix cases where NumPyPrinter CodePrinter generates code with incorrect precedence when rewriting functions.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* printing
  * Fix cases where NumPyPrinter CodePrinter generates code with incorrect precedence when rewriting functions.
<!-- END RELEASE NOTES -->

mgreminger committed:
fix: fix CodePrinter output for function rewrite

In some cases, rewriting the function results in an expression that is not a single function. In these cases, the expression should be wrapped in parenthesis to preserve the original order of operations when the function is part of a larger expression.

mgreminger committed:
tests: update tests for CodePrinter change

moorepants commented:
Nice catch here, thanks.

Your current implementation has the effect of just adding parentheses to the top level Add,Mul,Pow. Ideally, it wouldn't do that.

I'm ok with parentheses being added in all other time a function is rewritten though, as that ensures we don't miss any edge cases.

You should also add a new unit test with your example failing case above.

mgreminger committed:
fix: enclose all rewritten functions in parenthesis

This ensures that the resulting function will always have its original function call precedence even if it is composed of other operations.

mgreminger commented:
Thanks for the feedback, makes sense. I've updated to always enclose the rewritten function in parenthesis to preserve the precedence of the original function call. I've also added the test case from the example above to the tests suite.
